scripts/build_test.py

```diff
@@ -24,7 +33,79 @@
 class BuildTests(test_utils.GenericTestBase):
     """Test the build methods."""
 
+    def test_minify(self):
+        """Tests _minify with an invalid filepath."""
+        # pylint: disable=protected-access
+        with self.assertRaises(subprocess.CalledProcessError) as calledProcess:
+            build._minify(
+                'path/to/input.js', 'path/to/output.js')
+        # pylint: enable=protected-access
+        calledProcessException = calledProcess.exception
+        # returncode is the exit status of the child process.
+        self.assertEqual(calledProcessException.returncode, 1)
+
+    def test_join_files(self, file_count=10):
+        """Determine third_party.js contains the content of the first 10 JS
+        files in /third_party/static.
+        """
+        third_party_js = os.path.join(
+            build.THIRD_PARTY_GENERATED_DEV_DIR, 'js', 'third_party.js')
+        # Get all filepaths from manifest.json.
+        dependency_filepaths = build.get_dependencies_filepaths()
+        # Join all JS files in /third_party/static.
+        # pylint: disable=protected-access
+        build._join_files(dependency_filepaths['js'], third_party_js)
+        # pylint: enable=protected-access
+        with open(third_party_js, 'r') as joined_js_file:
```

Comment:
Would be better if tests didn't directly modify the filesystem. Maybe you can refactor the code so that this is a straightforward input/output test of a function, and pull out the filesystem modifying parts somewhere else.

An alternative is to mock out the os calls (cf. self.swap).

Reply:
Looks like I need to do some refactoring on various functions in `build.py` that handle file operation, the general direction would look something like this article:
http://undrivendevelopment.blogspot.com/2013/03/we-may-often-come-across-piece-of-code.html
Does this look good to you?
So basically I would be replacing : 
```Python
def _join_files(source_paths, target_file_path):
```
with 
```Python
def _join_files(source_paths, target_file_stream):
```
and call _join_files with an `with open(..)` wherever needs it (`build_third_party_lib()`)
Does that sound good? @seanlip

Reply:
Yes, something like that should work! Basically you don't want your tests to have side-effects on the env.

Reply:
There is a WIP to refactor all file I/O functions like`_join_files`. Thanks!

Reply:
WIP is over! `build_test` is ready for another review! @seanlip